Changes for jira PUBSTWO-1669 (Query by doi)

orosz-usgs opened the pull request:
Refactored MpPublicationMvcService, PwPublicationMvcService and PwPublicationRssMvcService to use
a pojo object for the query parameter in the method for a http GET
  Update MvcService.buildFilters to just take the pojo filtering class as the single argument
  Moved the current functionality of the "doi" parameter to a new "hasDoi" parameter
  Updated the functionality of the "doi" parameter match on the doi string value.
  Updated MvcServiceTest and PublicationDaoIT

Commit:
Changes for jira PUBSTWO-1669 (Query by doi)
  Refactored MpPublicationMvcService, PwPublicationMvcService and PwPublicationRssMvcService to use
a pojo object for the query parameter in the method for a http GET
  Update MvcService.buildFilters to just take the pojo filtering class as the single argument
  Moved the current functionality of the "doi" parameter to a new "hasDoi" parameter
  Updated the functionality of the "doi" parameter match on the doi string value.
  Updated MvcServiceTest and PublicationDaoIT

Commit:
Changes for jira PUBSTWO-1669 (Query by doi)
  Added BibndingException  to GlobalDefaultExceptionHandler sop that the user see back
a short summary of the format error in the query parameters:
   "Error in parameter chorus: Invalid boolean value [xyz]"

dsteinich commented on src/main/java/gov/usgs/cida/pubs/webservice/GlobalDefaultExceptionHandler.java:
```diff
@@ -59,4 +65,26 @@
+			String mess = err.getDefaultMessage() == null ? " Error processing field" : err.getDefaultMessage();
+			mess = mess.replaceAll(NESTED_EXCEPTION_REGEX, "");
+			String prefix = "Error in parameter " + err.getField() + ":";
+			errorMess = errorMess + sep + prefix + mess;
```
Would it be better to use a StringBuilder here (rather than concatenating).

dsteinich commented on src/main/java/gov/usgs/cida/pubs/webservice/mp/MpPublicationMvcService.java:
```diff
@@ -68,41 +66,25 @@ public MpPublicationMvcService(@Qualifier("publicationBusService")
 
 	@ApiOperation(value = "", authorizations = { @Authorization(value=PubsConstantsHelper.API_KEY_NAME) })
 	@GetMapping
+	@RequestMapping(method=RequestMethod.GET)
```
What is the reasoning behind using both `@GetMapping` and `@RequestMapping`?

orosz-usgs replied:
it was one of those that I could have swore it was needed. Do not know what was wrong
when I tried it earlier. Yes, the second annotation is not needed

dsteinich commented on src/main/java/gov/usgs/cida/pubs/webservice/mp/MpPublicationMvcService.java:
```diff
@@ -68,41 +66,25 @@ public MpPublicationMvcService(@Qualifier("publicationBusService")
-				indexId, ipdsId, listId, null, null, null, null, null, pageRowStart,
-				pageSize, prodId, pubAbstract, null, null, null, q, null, null, reportNumber,
-				seriesName, startYear, subtypeName, title, typeName, year);
+		System.out.println("DEBUG: filetr-= " + filterParams);
```
Please use LOG.debug rather than System.out.println

dsteinich commented on src/main/java/gov/usgs/cida/pubs/webservice/pw/PwPublicationMvcService.java:
```diff
@@ -79,55 +79,21 @@ public PwPublicationMvcService(
 
 	@GetMapping(produces={MediaType.APPLICATION_JSON_VALUE,
 			PubsConstantsHelper.MEDIA_TYPE_XLSX_VALUE, PubsConstantsHelper.MEDIA_TYPE_CSV_VALUE, PubsConstantsHelper.MEDIA_TYPE_TSV_VALUE})
+	@RequestMapping(method=RequestMethod.GET)
```
Same question on using both annotations.

dsteinich commented on src/main/java/gov/usgs/cida/pubs/webservice/pw/PwPublicationRssMvcService.java:
```diff
@@ -64,50 +66,19 @@ public PwPublicationRssMvcService(@Qualifier("pwPublicationBusService")
-			@RequestParam(value=PwPublicationDao.CHORUS, required=false) Boolean chorus,
-			HttpServletResponse response) {
-
+	@RequestMapping(method=RequestMethod.GET)
```
Same question on using both annotations.

dsteinich commented on src/main/resources/mybatis/mappers/publication.xml:
```diff
@@ -427,9 +427,17 @@
             </if>
 
-            <if test="null != doi">
+            <if test="doi != null and doi.length > 0">
```
the `(` and `)` could be moved to the `foreach` - `<foreach item="d" collection="doi" open="(" separator="or" close=")">`
Might even be able to move the `and` into the `open`

dsteinich replied:
I'd really like to see the parenthesis moved to the open and close.

dsteinich commented on src/test/java/gov/usgs/cida/pubs/dao/PublicationDaoIT.java:
```diff
@@ -355,7 +355,8 @@ public void getObjectCountTest() {
 		filters.put(PublicationDao.CONTRIBUTOR, "contributor1" + MvcService.TEXT_SEARCH_STARTS_WITH_SUFFIX + MvcService.TEXT_SEARCH_AND + "contributor2" + MvcService.TEXT_SEARCH_STARTS_WITH_SUFFIX);
 		
-		filters.put(PublicationDao.DOI, true);
+		filters.put(PublicationDao.DOI, new String[]{"DOI-123"});
```
Can you add a second doi here to show that the foreach is set up correctly? (The separator is valid.)

It also looks like the getByMapTest() in this class hasn't been truly kept up to date. There should be separate tests - one showing that specifying just DOI values returns the correct publications and one that shows specifying just the hasDoi returns the correct publications.  (Might also be time to split it into individual test.) 

Since PUBTSWO-1679 was also tagged for this pull request, orcid should also be tested here.

Commit:
Changes for jira PUBSTWO-1669 (Query by doi)
  Changes from review
  Added orcidid and doi tests
  Updated test data to use 19 digit ORCID
  Add two publications linked to contributors that have an orcid assigned

Commit:
Changes for jira PUBSTWO-1669 (Query by doi)
 Sql syntax clean up publication.xml
